File: HouseRobber.swift
// Time Complexity : n^m
// Space Complexity : n^m
// Did this code successfully run on Leetcode : No. Time limit exceeded
// Any problem you faced while coding this : mistakenly put the base case


// Your code here along with comments explaining your approach
//Recursive approach to choose and not choose a house then get the max of two cases 

class Solution {
    func rob(_ nums: [Int]) -> Int {
        if nums.count == 0 {
            return 0
        }
        return helper(nums,0,0)
    }
    
    private func helper(_ nums: [Int], _ startIndex: Int, _ amount: Int) -> Int {
        //base
        if startIndex >= nums.count {
            return amount
        }
        
        //logic
        //choose a coin
        let case2 = helper(nums, startIndex+2, nums[startIndex] + amount)
        //not choose a coin
        let case1 = helper(nums, startIndex+1, amount)
        
        //result
        return max(case1,case2)
    }
}

//DP soulution
//previous

class Solution {
    func rob(_ nums: [Int]) -> Int {
        if nums.count == 0 {
            return 0
        }
        var choose = 0
        var notChoose = 0
        var prevChoosen = 0
        for i in 0..<nums.count {
           //choose
            choose = notChoose + nums[i]
           //last not choosen + current house
           //not choose
            notChoose = max(notChoose,prevChoosen)
           // last maximum of choose and not choose
            prevChoosen = choose
        }
        
        return max(choose, notChoose)
    }

}

File: HouseRobber.js
// Time Complexity : O (n)
// Space Complexity : O (1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach

/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function(nums) {
  if (nums.length === 0 || nums === null) {
    return 0;
  }

  var n = nums.length;

  // using n * 2 matrix
  // var dp = new Array(2);
  // for (var i = 0; i <= n; i++) {
  //   dp[i] = new Array(n);
  // }

  // dp[0][0] = 0; // default
  // dp[0][1] = nums[0];

  // for (var i = 1; i < n; i++) {
  //   dp[i][0] = Math.max(dp[i - 1][0], dp[i - 1][1]);
  //   dp[i][1] = dp[i - 1][0] + nums[i];
  // }

  // No extra space
  var skip = 0,
    take;
  take = nums[0];
  for (var i = 1; i < n; i++) {
    var temp = skip;
    skip = Math.max(skip, take);
    take = temp + nums[i];
  }

  return Math.max(skip, take);
};
